feat: faster asset deployment

bestickley opened the pull request:
Fixes https://github.com/jetbridge/cdk-nextjs/issues/117, https://github.com/jetbridge/cdk-nextjs/issues/113

Updates:
Refactor NextjsAssetDeployment into NextjsStaticAssets. The static assets bucket now lives in this construct so the rename semantically makes the most sense.
Refactor NextjsS3EnvRewriter to NextjsBucketDeployment. This construct now doesn't just rewrite code, it copies it to a new location. New name was inspired by CDK's BucketDeployment but we needed additional capabilities hence needing to create our own.
Add top level prop skipBuild to re-deploy without running open-next build
Only run open-next build when it's going to be deployed
Increase memory on asset deployment function (10x increase in time savings!)
Move bundling of lambda functions that don't depend upon user code into the building of this package for faster user deployment.
Remove archive functionality as it's already by CDK Asset (see more [here](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3_assets-readme.html#how-does-it-work))
Remove dependencies that aren't needed anymore: aws-sdk, jszip, cross-spawn
Remove cacheBucket from NextjsDefaultProps which should have been removed before.
Remove NextjsProps.isPlaceholder. It is useful when deploying this construct with SST, but if you're using SST, you should be using NextjsSite. Therefore, isPlaceholder isn't needed in this construct.
Remove NextjsBuild.projectRoot as it's not used for anything.
Remove NextjsLayer as it's not used by Nextjs
Reason Why We Cannot use NodejsFunction for all Lambdas
Originally I wanted to use NodejsFunction which manages bundling for us. However, we cannot couple bundling and deploying the function together when we don't know deploy time values because this would deploy the function without inject environment variables. The period of the function being in this state would be small, but isn't acceptable IMO. Therefore, we must deploy assets to S3, inject environment variables, then deploy the function - similar to how it works currently. There are still improvements to make, though. Server and Image functions must be inject with env vars where as Revalidation and in the future, Warmer can be bundled with NodejsLambda.

bestickley commented:
Reviewers, please follow instructions in examples/README to run e2e tests to verify correct behavior. I'm still working on one issue with the HOST header with @khuezy that I need to resolve before merging.

bestickley commented:
@kevin-mitchell, looks like you're not a part of this organization so I cannot add you as an official reviewer but would appreciate your review as well!

kevin-mitchell commented on src/NextjsImage.ts:
```diff
@@ -35,56 +26,37 @@ export interface NextjsImageProps extends NextjsBaseProps {
-        statements: [policyStatement],
-      })
-    );
+    bucket.grantRead(this);
```
This is nice, I was confused previously why we had more complex policy definitions. There may be good reasons for them of course, but if we have no specific reason the `bucket.grandRead(this)` syntax is really easy to read and (I think?) a big reason people like CDK.

kevin-mitchell commented:
Wow, this is amazing @bestickley! A huge amount of work but looks like a huge step forward in terms of removing cruft and improving the general understandability (possibly not a word) of the project. Thank you! 

I took a high level look at the changes but haven't really internally processed the details at this point. I'm in the middle of a move so am sort of not in a great headspace for deeply understanding stuff quickly. THAT SAID, I did try to follow the example README (btw I think this is really nice to have for a place to sort of sanity check themselves), and ran into build issues.

```
Failed to compile.

./app/ssr/page.tsx
Module not found: Can't resolve '@open-next/utils'
```

This is probably user error or something with my local setup, looks like something I've run into in the past, but I _think_ I fairly exactly followed the README directions so figured I'd mention it.  

I'm going to try to take a closer / better look and get the example working if I can tomorrow. For now, this looks great and again thank you so much for the huge effort!

bestickley commented:
Thank you, @kevin-mitchell! I appreciate your time reviewing.
Sorry about that, I missed a step in the README. I've since updated it. You need to run `pnpm build` within `open-next` git submodule.

revmischa commented on README.md:
```diff
@@ -10,20 +10,25 @@ Supported NextJs versions: >=12.3.0+ (includes 13.0.0+)
+  constructor(scope: Construct, id: string, props?: StackProps) {
+    super(scope, id, props);
+    const nextjs = new Nextjs(this, 'Nextjs', {
+      nextjsPath: '../../web', // relative path to Next.js project root
```
why would this be `../../web`?
isn't it the nextjs project root relative to the repo root?

bestickley replied:
@revmischa, thanks for your review!
It's just an example. I arbitrarily chose `../../web`. Would you prefer `../web`?
Yes, " the nextjs project root relative to the repo root". Nothing has change with `nextjsPath`.

revmischa replied:
```suggestion
      nextjsPath: './web', // relative path from your project root to NextJS
```

kevin-mitchell commented:
> Thank you, @kevin-mitchell! I appreciate your time reviewing. Sorry about that, I missed a step in the README. I've since updated it. You need to run `pnpm build` within `open-next` git submodule.

I did something dumb, which is I checked out a fresh copy of the project, followed the readme example, but I didn't actually build the main project (i.e. `cdk-nextjs-standalone`) before installing in the specific example. I then built the main project,reinstalled everything in the example (though worth noting this has to be forced I think?)

All of this is to say, after building the project the `app-router` worked / deployed!

I guess my only take away here is it MIGHT be worth mentioning in the examples readme that the actual project needs to be built before running `pnpm install` with the example dir. Or maybe that'll be very obvious to anybody checking out this project. 

---

I also have a question / comment re: `## PNPM Monorepo Symlinks`

I use `npm` 100% of the time if possible. If I absolutely must use `pnpm` or `yarn` or anything else for some reason I will, but I try not to. That said, I don't fully "get" this part of the readme to be honest. It reads like there is some extra complexity because of `pnpm`? Is this only relevant to those who choose to use `pnpm`?

bestickley commented:
@kevin-mitchell,
Definitely not dumb. I added a "Prerequisites" section in examples/README.md ensuring reader builds cdk-nextjs-standalone.
Good question on PNPM Monorepo Symlinks. I added an italicized section below it that states it's only applicable is using PNPM. I use PNPM anytime I can so I wanted to call it out. It's more of a note to future developers, not for typical users of construct.
